### PR TITLE
Add BioBlend tests

### PR DESCRIPTION
**Pull request by arash77**
This pull request adds BioBlend tests to the project.
https://github.com/galaxyproject/bioblend/blob/main/.github/workflows/test.yaml

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: .github/workflows/bioblend.yaml [hunk 1]
```diff
@@ -0,0 +1,128 @@
+name: Tests
+on:
+  pull_request:
+    paths:
+      - "lib/galaxy/webapps/galaxy/api/**"
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]:
```suggestion
      - .github/workflows/bioblend.yaml
      - lib/galaxy/schema/**
      - lib/galaxy/webapps/galaxy/api/**
      - lib/galaxy/webapps/galaxy/services/**
```

File: .github/workflows/bioblend.yaml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  workflow_dispatch:
+
+concurrency:
+  group: test-${{ github.ref }}
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref }}
```

File: .github/workflows/bioblend.yaml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  cancel-in-progress: true
+jobs:
+  test:
+    runs-on: ${{ matrix.os }}
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/bioblend.yaml [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        run: |
+          # Create a PostgreSQL database for Galaxy. The default SQLite3 database makes test fail randomly because of "database locked" error.
+          createdb -U postgres galaxy
+          # Run ToolShed tests only once per Python version
+          if [ "${{ matrix.galaxy_version }}" = 'dev' ]; then
+              export BIOBLEND_TOOLSHED_URL=https://testtoolshed.g2.bx.psu.edu/
+          fi
+          # Install Galaxy
+          GALAXY_DIR=galaxy-${{ matrix.galaxy_version }}
+          git clone --depth=1 -b ${{ matrix.galaxy_version }} https://github.com/galaxyproject/galaxy $GALAXY_DIR
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]: As mentioned, get galaxy using `actions/checkout`.

```suggestion
```

File: .github/workflows/bioblend.yaml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+          - release_19.05
+        include:
+          - os: ubuntu-latest
+            tox_env: py313
+            galaxy_version: dev
+          # Cannot test on macOS because service containers are not supported
+          # yet: https://github.community/t/github-actions-services-available-on-others-vms/16916
+          # - os: macos-latest
+          #   tox_env: py38
+          #   galaxy_version: dev
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]: You don't need this big matrix. Drop `os` and `galaxy_version` (you should use `actions/checkout` to get galaxy at the commit being tested), and choose a Python version for Galaxy (I'd say '3.8') and one for BioBlend ('3.13') to test on.

Comment (arash77) [reply .github/workflows/bioblend.yaml]: What about the `tox_env`, as it determines the BioBlend python version?

Comment (nsoranzo) [reply .github/workflows/bioblend.yaml]: You can keep `tox_env: [py313]`, as the BioBlend Python version is then derived from that already in the steps below.

Comment (nsoranzo) [reply .github/workflows/bioblend.yaml]: (Amended the above comments to say Python 3.13 for BioBlend).

File: .github/workflows/bioblend.yaml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+              galaxy_python_version=3.7
+              ;;
+            release_22.0? | release_23.? )
+              galaxy_python_version=3.7
+              ;;
+            release_24.? | dev )
+              galaxy_python_version=3.8
+              ;;
+          esac
+          echo "galaxy_python_version=$galaxy_python_version" >> $GITHUB_OUTPUT
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]: Drop this in favour of specifying `galaxy_python_version: ['3.8']` in the job matrix.

**Comment by arash77** [comment]
Thank you for the suggestions

File: .github/workflows/bioblend.yaml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          python-version: ${{ matrix.galaxy_python_version }}
+      - name: Run tests
+        env:
+          PGPASSWORD: postgres
+          PGPORT: 5432
+          PGHOST: localhost
+        run: |
+          # Create a PostgreSQL database for Galaxy. The default SQLite3 database makes test fail randomly because of "database locked" error.
+          createdb -U postgres galaxy
+          export BIOBLEND_TOOLSHED_URL=https://testtoolshed.g2.bx.psu.edu/
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]: No need to test BioBlend's interactions with the Test ToolShed.

```suggestion
```

File: .github/workflows/bioblend.yaml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        id: get_bioblend_python_version
+        run: echo "bioblend_python_version=$(echo "${{ matrix.tox_env }}" | sed -e 's/^py\([3-9]\)\([0-9]\+\)/\1.\2/')" >> $GITHUB_OUTPUT
+      - name: Set up Python for BioBlend
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ steps.get_bioblend_python_version.outputs.bioblend_python_version }}
+      - name: Install tox
+        run: |
+          python3 -m pip install --upgrade pip setuptools
+          python3 -m pip install 'tox>=1.8.0' 'virtualenv>=20.0.14'
```
Comment (nsoranzo) [inline .github/workflows/bioblend.yaml]: virtualenv was only needed by old Galaxy versions.

```suggestion
          python3 -m pip install 'tox>=1.8.0'
```

**Review by nsoranzo — APPROVED** [review]
19 BioBlend tests fail as expected, thanks!
I'd say we merge this and make the tests pass again in a separate PR.

**Comment by nsoranzo** [comment]
Thanks @arash77 !

**Pull request merged**